fmriprep/utils/misc.py

```diff
@@ -63,38 +61,44 @@ def collect_bids_data(dataset, subject, spec=None, session=None, run=None):
         't1w': {'type': 'T1w', 'ext': 'nii'}
     }
 
-    #  Add a subject key pair to each query we make so that we only deal with
-    #  files related to this workflows specific subject. Could be made opt...
-    for key in queries.keys():
-        queries[key]['subject'] = subject
-
-    imaging_data = copy.deepcopy(INPUTS_SPEC)
-    fieldmap_files = [x.filename for x in layout.get(**queries['fmap'])]
-    imaging_data['fmap'] = fieldmap_files
-    t1_files = [x.filename for x in layout.get(**queries['t1w'])]
-    imaging_data['t1w'] = t1_files
-    sbref_files = [x.filename for x in layout.get(**queries['sbref'])]
-    imaging_data['sbref'] = sbref_files
-
-    loop_on = ['session', 'run', 'acquisition', 'task']
-    get_kwargs = {}
-
-    for key in loop_on:
-        unique_list = layout.unique(key)
-        if unique_list:
-            get_kwargs[key] = unique_list
-
-    query_kwargs = []
-    for key in get_kwargs:
-        query_kwargs.append([(key, x) for x in get_kwargs[key]])
-
-    query_kwargs = itertools.product(*query_kwargs)
-
-    for elem in query_kwargs:
-        epi_files = [x.filename for x
-                     in layout.get(**dict(dict(elem), **queries['epi']))]
-        if epi_files:
-            imaging_data['func'].append(epi_files)
+    for session in session_list:
+        imaging_data = {}
+
+        #  Add a subject key pair to each query we make so that we only deal with
+        #  files related to this workflows specific subject. Could be made opt...
+        for key in queries.keys():
+            queries[key]['subject'] = subject
+
+        session_data = copy.deepcopy(INPUTS_SPEC)
+        fieldmap_files = [x.filename for x in layout.get(**queries['fmap'])]
+        session_data['fmap'] = fieldmap_files
+        t1_files = [x.filename for x in layout.get(**queries['t1w'])]
+        session_data['t1w'] = t1_files
+        sbref_files = [x.filename for x in layout.get(**queries['sbref'])]
+        session_data['sbref'] = sbref_files
+
+        loop_on = ['session', 'run', 'acquisition', 'task']
```

Comment:
I think @rwblair was serializing sessions as well with this loop. I think this file deserves a different PR where he can check it out and you can go through it together.